Reving up package for System.Runtime

joperezr opened the pull request:
cc: @ericstj @weshaggard @danmosemsft 

This are the required changes in order to update the package contract and implementation. With this approach, we are cross-compiling the src for netstandard1.5, but another one could be to simply use a depproj for older implementations(like netstandard1.5). Thoughts?

joperezr commented:
Once we are ok with this approach, I'll send out a PR that basically replicates this across all of the contracts we've touched in dev/api branch.

joperezr commented:
related: #10298

joperezr commented:
@dotnet-bot test Innerloop Windows_NT Release Build and Test please

joperezr commented:
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test please

joperezr commented:
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test please

joperezr commented:
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test please

joperezr commented:
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test please

joperezr commented:
@dotnet-bot Innerloop Ubuntu14.04 Release Build and Test please

joperezr commented:
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test please

weshaggard commented on src/System.Runtime/src/System.Runtime.csproj:
```diff
@@ -4,35 +4,42 @@
+  <ItemGroup Condition="'$(TargetGroup)' != 'net462' And '$(TargetGroup)' != 'net463'">
     <ProjectReference Include="..\..\System.Private.Uri\src\System.Private.Uri.csproj">
       <OSGroup>Windows_NT</OSGroup>
+      <UndefineProperties>TargetGroup</UndefineProperties>
```
I assume this is to prevent duplicated versions of this library build and we only need the reference so it doesn't matter which we get. For completeness can you add `%(ProjectReference.UndefineProperties);TargetGroup`

joperezr replied:
Will do

weshaggard commented on src/System.Runtime/src/project.json:
```diff
@@ -1,12 +1,15 @@
+      }
+    },
     "netstandard1.5": {
+      "__comment": "This should be a stable version of the TargetingPack package that matches netcoreapp1.0 RTM",
```
Any reason we cannot make that change now?

joperezr replied:
I believe we didn't actually ship a stable version of the Targeting package, so the work would be to select one that matches netcoreapp and re-publish it with a stable version on it, and then add it here. And no, there is no reason why we can't make that now, just wanted to get this PR out first to get feedback.

joperezr commented:
@dotnet-bot test Innerloop Ubuntu14.04 Release Build and Test please

joperezr commented on src/System.Collections/tests/System.Collections.Tests.csproj:
```diff
@@ -25,7 +25,7 @@
       <OSGroup>$(InputOSGroup)</OSGroup>
     </ProjectReference>
-    <ProjectReference Include="..\..\System.Runtime\pkg\System.Runtime.pkgproj">
+    <ProjectReference Include="..\..\System.Runtime\src\System.Runtime.csproj">
```
We need to do this because the src/System.Collections.csproj references the live csproj as well, which means that System.Collections.dll now depends on 4.2.0.0 version of System.Runtime.dll. Since we use to have a pkgproj reference in the tests and we are testing against netcoreapp1.0, the resolving System.Runtime for the tests was the one for netstandard1.5 (version 4.1.0.0) which is why the tests were failing. We will be able to remove this once System.Runtime package version is rev'ed and we update the corefx prerelease dependencies.

weshaggard replied:
As we chatted offline lets try to update the ProjectReference in System.Collections to get the netstandard1.5 version instead of this change.

joperezr commented on src/System.Private.Uri/tests/ExtendedFuctionalTests/System.Private.Uri.ExtendedFunctional.Tests.csproj:
```diff
@@ -7,6 +7,7 @@
     <OutputType>Library</OutputType>
     <AssemblyName>System.Private.Uri.Tests</AssemblyName>
     <NugetTargetMoniker>.NETStandard,Version=v1.5</NugetTargetMoniker>
+    <NoWarn>$(NoWarn);1685</NoWarn>
```
This warning is because we added mscorlib targetting pack reference, so now Task<> exists in both mscorlib and System.Runtime. This will work for now, but we have to eventually remove this and make the project target TestTFM netcoreapp1.1 to get the right assets out of System.Runtime.pkgproj

joperezr commented:
FYI: @sepidehMS I had to make some changes on the extendedFunctional tests in order for my update to work.

joperezr commented:
I have addressed all the feedback and I would like to get this in soon since I want people adding new tests to start working with the right version of the package. Please let me know if there are any additional comments

joperezr commented:
ping

ericstj commented on src/System.Private.Uri/tests/ExtendedFuctionalTests/System.Private.Uri.ExtendedFunctional.Tests.csproj:
```diff
@@ -2,11 +2,13 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
+    <TestTFM>netcoreapp1.1</TestTFM>
+    <FilterToTestTFM>netcoreapp1.1</FilterToTestTFM>
```
I thought these were properties that were supposed to be passed in rather than directly authored in projects?

joperezr replied:
This project will only work in netcoreapp1.1 which is why I added them here, but I can also add it on the builds file if you think its best

ericstj commented on src/System.Runtime/ref/4.1.0/project.json:
```diff
@@ -0,0 +1,12 @@
+  },
+  "frameworks": {
+    "netstandard1.5": {
+      "imports": [
```
You can remove this import

joperezr replied:
Will do.

ericstj commented on src/System.Private.Uri/tests/ExtendedFuctionalTests/System.Private.Uri.ExtendedFunctional.Tests.csproj:
```diff
@@ -20,6 +22,7 @@
     </Compile>
   </ItemGroup>
   <ItemGroup>
+    <TestNugetTargetMoniker Include=".NETStandard,Version=v1.7" />
```
This doesn't look right /cc @karajas.

joperezr replied:
This is a recently added project which tests purely new api, so it is intended only to work on netcoreapp1.1 or higher.

ericstj commented on src/System.Runtime/tests/System.Runtime.Tests.csproj:
```diff
@@ -12,6 +12,7 @@
     <NoWarn>1718</NoWarn>
     <IncludePerformanceTests>true</IncludePerformanceTests>
     <NugetTargetMoniker>.NETStandard,Version=v1.5</NugetTargetMoniker>
+    <NugetTargetMoniker Condition="'$(TestTFMs)'=='netcoreapp1.1'">.NETStandard,Version=v1.7</NugetTargetMoniker>
```
Shouldn't NuGetTargetMoniker already be set by the targetgroup?  If so, then just add a condition to the line above to either only set it for empty targetgroup or only set it if it isn't already set.

ericstj commented on src/System.Runtime/tests/System.Runtime.Tests.csproj:
```diff
@@ -107,6 +108,7 @@
     <Compile Include="$(CommonTestPath)\System\PerfUtils.cs">
       <Link>Common\System\PerfUtils.cs</Link>
     </Compile>
+    <Compile Condition="'$(TestTFMs)'=='netcoreapp1.1'" Include="netstandard.1.7.Tests.cs" />
```
Normally we use TargetGroup for compile conditions.

joperezr commented:
@ericstj thanks for the feedback, I've addressed it on 5ccb3ed let me know what you think.

ericstj commented:
LGTM

weshaggard commented on src/System.Private.Uri/tests/System.Private.Uri.Tests.builds:
```diff
@@ -16,6 +16,16 @@
     </Project>
+    <Project Include="ExtendedFunctionalTests\System.Private.Uri.ExtendedFunctional.Tests.csproj">
+      <TargetGroup>netstandard1.7</TargetGroup>
+      <SkipTestRun>true</SkipTestRun>
```
Why do we need to explicitly do the skipping here?

weshaggard replied:
The skipping should be handled by setting the TestTFMs to be netcoreapp1.1

joperezr replied:
Yup, changed.

---
In reply to: [72678764](https://github.com/dotnet/corefx/pull/10297#discussion_r72678764) [](ancestors = 72678764)

weshaggard commented on src/System.Private.Uri/tests/System.Private.Uri.Tests.builds:
```diff
@@ -16,6 +16,16 @@
       <OSGroup>Windows_NT</OSGroup>
     </Project>
+    <Project Include="ExtendedFunctionalTests\System.Private.Uri.ExtendedFunctional.Tests.csproj">
+      <TargetGroup>netstandard1.7</TargetGroup>
```
Shouldn't this target be netcoreapp1.1?

joperezr replied:
Discussed offline. We agreed that netstandard1.7 is the right TargetGroup

---
In reply to: [72678399](https://github.com/dotnet/corefx/pull/10297#discussion_r72678399) [](ancestors = 72678399)

weshaggard commented on dir.traversal.targets:
```diff
@@ -122,7 +122,7 @@
 
       <!-- for projects not defining TestTFM in .builds files, we assume the TestTFM is the default which is netcoreapp1.0 -->
-      <ProjectsToTest Include="@(Project)" Condition="'%(Project.Extension)'=='.csproj' And '$(FilterToTestTFM)' == 'netcoreapp1.0' And '%(Project.TestTFMs)' == ''" />
+      <ProjectsToTest Include="@(Project)" Condition="'%(Project.Extension)'=='.csproj' And '$(FilterToTestTFM)' == 'netcoreapp1.0' And '%(Project.TestTFMs)' == '' And '%(Project.SkipTestRun)'!='true'" />
```
We shouldn't need this explicit skip property.

joperezr replied:
Removed

---
In reply to: [72678855](https://github.com/dotnet/corefx/pull/10297#discussion_r72678855) [](ancestors = 72678855)

weshaggard commented on src/System.Runtime/tests/System.Runtime.Tests.builds:
```diff
@@ -4,6 +4,10 @@
     <Project Include="System.Runtime.Tests.csproj"/>
     <Project Include="System.Runtime.Tests.csproj">
+    	<TargetGroup>netstandard1.7</TargetGroup>
+    	<SkipTestRun>true</SkipTestRun>
```
TestTFMs netcoreapp1.1

joperezr replied:
yes, added.

weshaggard commented on src/System.Runtime/tests/System.Runtime.Tests.csproj:
```diff
@@ -107,6 +107,7 @@
     <Compile Include="$(CommonTestPath)\System\PerfUtils.cs">
       <Link>Common\System\PerfUtils.cs</Link>
     </Compile>
+    <Compile Condition="'$(TargetGroup)'=='netstandard1.7'" Include="netstandard.1.7.Tests.cs" />
```
This is fine for now but it won't scale. Eventually we will want to add more tests and at that point we should move the condition to an item group and name the files based on what they are testing.

joperezr replied:
Agreed, we will have to fix this eventually.

---
In reply to: [72679281](https://github.com/dotnet/corefx/pull/10297#discussion_r72679281) [](ancestors = 72679281)

weshaggard commented on src/System.Private.Uri/tests/ExtendedFunctionalTests/System.Private.Uri.ExtendedFunctional.Tests.csproj:
```diff
@@ -6,7 +6,7 @@
     <ProjectGuid>{0febe054-68ac-446f-b999-9068736d3cec}</ProjectGuid>
     <OutputType>Library</OutputType>
     <AssemblyName>System.Private.Uri.Tests</AssemblyName>
-    <NugetTargetMoniker>.NETStandard,Version=v1.5</NugetTargetMoniker>
```
Are all the tests in here new and only useful for netstandard1.7 stuff? If so I'm good with this change if not then we need to correctly split the tests.

joperezr replied:
Yes, all tests in this project are for netstandard1.7 Apis

weshaggard commented on src/System.Runtime/tests/System.Runtime.Tests.builds:
```diff
@@ -2,7 +2,12 @@
+    <Project Include="System.Runtime.Tests.csproj">
+      <TargetGroup>netstandard1.5</TargetGroup>
+    </Project>
+    <Project Include="System.Runtime.Tests.csproj">
```
Something doesn't quite look right here. Shouldn't the one entry be empty as before and the new entry be netstandard1.7?

joperezr replied:
The default is now netstandard1.7 so that is the one without targetgroup, and the one with TargetGroup is the old one

joperezr replied:
actually I believe you are right, I didn't change the default here, let I'll go and fix that now.

weshaggard replied:
Lets not change the default to ns17 yet until we actually support netcoreapp1.1 runs at least.

joperezr replied:
ok, I've fixed the builds file now without defaulting to 1.7

joperezr commented:
@dotnet-bot test Interloop CentOS7.1 Release Build and Test please

joperezr commented:
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test please
